Document PE Bounty and Bond Design (#702)
omgnetwork/plasma-contracts: plasma_framework/docs/design/exit_bond_and_bounty_mechanics.md

# Exit Bond and Bounty Mechanics

This document is a description of the current implementation of exit bonds and process exit bounties which are an integral part of the exit mechanism.

A Process Exit Bounty is a reward which is paid out to the `processExit()` initiator to compensate for the transaction cost. The motive of having a bounty is to remove an extra step of processing exits while exiting funds. This also ensures everyone pays their share of processing contrary to passing on the responsibility to someone else.

The Process Exit Bounty is paid out as a portion (or whole) of the Exit Bond and the size of the Bond is designed to ideally cover both the cost of challenging an invalid exit or the cost of processing a valid exit 

## The Bond Life Cycle

### For Standard Exits

> The Bond is supplied by the Output owner while starting an exit and the main motive is to disincentivize invalid exits while also being able to use a portion of the bond to incentivize processing exits. The Bond is internally divided into two portions. For successful exits, one of them can be returned to the output owner, and the other is a reward for processing the exit.

Reviewer: ```suggestion
The **Exit Bond** is supplied by the **Exit Initiator** when starting an exit. Its main motive is to disincentivize invalid exits, but in the case of a valid exit a portion of it is used to incentivize processing the exit by rewarding the **Exit Processor**.
```

> The initiator/owner has to obtaint the correct size of the Exit Bond by calling `startStandardExitBondSize()`.

Reviewer: ```suggestion
The **Exit Initiator** should obtain the correct size of the Exit Bond by calling `startStandardExitBondSize()`.
```

**_After Starting an Exit:_**

> *Bond -> from `Exit owner` to `Exit Game Contract`*

Reviewer: ```suggestion
*Bond -> from `Exit Initiator` to `Exit Game Contract`*
```

For an invalid exit, a challenge will transfer the whole bond to the challenger. Both exitable and non-exitable exits still remain in the exit queue and is deleted when the exits are processed. For a scenario where there are mutliple invalid exits in the queue, the processor is still awarded with some gas reward for deleting non-exitable exits from storage.

**_On succesfull challenge:_**
*Bond -> from `Exit Game Contract` to `Challenger`*

> The user who processes an exit, gets the bounty reward for processing an exit. The cost of processing multiple exits from the exit queue at once could be of less cost than processing all those exits individually. However, the bounty amount associated with each exit is given out to the processor.

Reviewer: ```suggestion
The **Exit Processor** gets the bounty as a reward for processing an exit. The cost of processing multiple exits from the exit queue at once is less than the cost of processing all those exits individually. However, the **Process Exit Bounty** amount associated with each exit is given out to the **Exit Processor**.
```

**_On processing exit:_**

> *Bond - Bounty -> from `Exit Game Contract` to `Exit owner`*\

Reviewer: ```suggestion
*Bond - Bounty -> from `Exit Game Contract` to `Exit Initiator`*\
```

*Bounty -> from `Exit Game Contract` to `Exit processor`*

> ### For In-Flght Exits

Reviewer: ```suggestion
### For In-Flight Exits
```

> The Bond/Bounty collection for in-flight exits is similar. However, for an in-flight exit, the Exit Bounty is a poriton of the Piggyback Bond. Since only the piggybacked input/output from an in-flight exit can be exited, it only makes sense to have the bounty for each piggyback from the respective owner. The size of the Piggyback Bond can be obtained by calling `piggybackBondSize()`.

Reviewer: ```suggestion
The situation for in-flight exits is slightly different. Since the piggybacked inputs or outputs of an in-flight exit are exited separately, the **Exit Initiator** is the owner of the piggybacked input or output. In this case the **Exit Bond** is the Piggyback Bond that is put up when piggybacking and the **Process Exit Bounty** is taken from this. The size of the Piggyback Bond can be obtained by calling `piggybackBondSize()`.
```

**_After Piggybacking an input/output:_**

> *Piggyback Bond -> from `I/O Owner` to `Exit Game Contract`*

Reviewer: ```suggestion
*Piggyback Bond -> from `Exit Initiator` to `Exit Game Contract`*
```

For any Input/Output spent challenge to the exit, the whole piggyback bond is transferred to the challenger. 

**_On succesfull challenge:_**
*Piggyback Bonds -> from `Exit Game Contract` to `Challenger`*

> Both the standard and in-flight exit land on the same exit queue and the process exit bounty associated with each exit (or specifically each input/output) is awarded to the processor.

Reviewer: ```suggestion
Both standard and in-flight exits are processed from the same exit queue and the **Process Exit Bounty** associated with each exit (or specifically each input/output) is awarded to the **Exit Processor**.
```

**_On processing exit:_**

> *Bond - Bounty -> from `Exit Game Contract` to `I/O Owner`*\

Reviewer: ```suggestion
*Bond - Bounty -> from `Exit Game Contract` to `Exit Initiator`*\
```

#### Standard Exits

The current size of the Exit Bond can be retrieved by

```
function startStandardExitBondSize() public view returns (uint128) {
        return startStandardExitBond.bondSize();
    }
```

The Exit Bond Size along with the portion for Exit Bounty can be updated by the maintainer, as long as it is within the updatable bounds pre-specified. The updated bond and bounty size will only be effective after the `effectiveUpdateTime`.

```
function updateStartStandardExitBondSize(uint128 newBondSize, uint128 newExitBountySize) public onlyFrom(framework.getMaintainer()) {
        startStandardExitBond.updateBondSize(newBondSize, newExitBountySize);
        emit StandardExitBondUpdated(newBondSize);
    }
```

A thing to note here is that the `newBondSize` should always be greater than or equal to the `newExitBountySize`. For situations where the Exit Bounty size has to be updated to an amount which is higher than the existing Exit Bond size, update the Exit Bond size accordingly.
For cases, when you want to update only one of the two entities, pass in the previous size for the one you don't want to change.

#### In-Flight Exits

> The current size of the In-flight Exit Bond can be retreived by

Reviewer: ```suggestion
The current size of the In-flight Exit Bond can be retrieved by
```

1. Alice starts a SE with B
2. The maintainer updates Bond size to B' and Bounty size to C'
3. Bob starts a SE within two days with B
4. Processor processes exits from queue and gets C+C as a reward
5. Alice and Bob both get back (B - C) each

##### Exit Bond is updated between two exits, but started after two days

1. Alice starts a SE with B
2. The maintainer updates Bond size to B' and Bounty size to C'
3. Bob starts a SE after two days with B'
4. Processor processes exits from queue and gets C+C' as a reward
5. Alice gets back (B - C) and Bob gets back (B' - C')

##### Exit Bond is updated again within effective period

1. Alice starts a SE with B
2. The maintainer updates PE Bounty to B' and Bounty size to C'
3. After one day, the maintainer updates the Bond size again to B" and Bounty size to C"
4. Then, after one day, Bob starts an SE with B
5. Processor processes exits from queue and gets C+C as a reward
6. Alice and Bob both get back (B - C) each

*Similarly for In-flight exits:*

##### Exit Bond is updated between Piggybacks

1. Alice spends an UTXO to Bob and Malorie
2. The exit is in-flight and Alice starts an IFE

> 3. Bob piggybacks the output with B

Reviewer: ```suggestion
3. Bob piggybacks an output with B
```

4. Then, the Piggyback Bond is updated to B' and Bounty to C'

> 5. After two days, Malorie piggybacks an output with B'

Reviewer: ```suggestion
5. After two days, Malorie piggybacks a different output with B'
```

# Exit Bond and Bounty Mechanics

This document is a description of the current implementation of exit bonds and process exit bounties which are an integral part of the exit mechanism.

> A Process Exit Bounty is a reward which is paid out to the `processExit()` initiator in return of the transaction cost borne by the person. The motive of having a bounty is to remove an extra step of processing exits while exiting funds. This also ensures everyone pays their share of processing contrary to passing on the responsibility to someone else.

Reviewer 2: ```suggestion
A Process Exit Bounty is a reward which is paid out to the `processExit()` initiator to compensate for the transaction cost . The motive of having a bounty is to remove an extra step of processing exits while exiting funds. This also ensures everyone pays their share of processing contrary to passing on the responsibility to someone else.
```

> The Exit Bounty is paid out as a portion (or whole) of the Exit Bond and the size of the Bond is designed to ideally cover both the cost of challenging an invalid exit or the cost of processing a valid exit 

Reviewer 2: ```suggestion
The Process Exit Bounty is paid out as a portion (or whole) of the Exit Bond and the size of the Bond is designed to ideally cover both the cost of challenging an invalid exit or the cost of processing a valid exit 
```

Reviewer 2: There is the `Process Exit Bounty` and `Exit Bond` do we need/use an extra term that is the sum of those values?

Author: Should we do that? Since we are providing a new value for the Bond Size instead of summing bond and bounty

Reviewer: I don't think we need an extra term, let's just stick to **Exit Bond** everywhere

## The Bond Life Cycle

### For Standard Exits

> The Bond is supplied by the Output owner while starting an exit and the main motive is to disincentivize invalid exits while also being able to use a portion of the bond to incentivize processing exits. The Bond is internally disintegrated into two portions. For successful exits, one of them can be returned to the output owner, and the other is a reward for processing the exit.

Reviewer 2: ```suggestion
The Bond is supplied by the Output owner while starting an exit and the main motive is to disincentivize invalid exits while also being able to use a portion of the bond to incentivize processing exits. The Bond is internally divided into two portions. For successful exits, one of them can be returned to the output owner, and the other is a reward for processing the exit.
```

> The initiator/owner has to obtaint the correct size of the Exit Bond by calling `startStandardExitBondSize()`.

Reviewer 2: Does it make sense to pick a terminology for the roles of the "exiting user" and "process exit initiator"?

Author: sure, any suggestions?

**_After Starting an Exit:_**
*Bond -> from `Exit owner` to `Exit Game Contract`*

> For an invalid exit, a challenge will transfer the whole bond to the challenger. Both exitable and non-exitable exits still remain in the exit queue and is deleted when the exits are processed. For a scenario where there are mutliple invalid exits in the queue, the processor is still awarded with some gas reward for deleting non-exitable exits from storage.

Reviewer 2: `Both exitable and non-exitable exits still remain in the exit queue and is deleted when the exits are processed.`
Shouldn't invalid exits be always non-exitable ?

Author: yes 👍, i'll rephrase this, wanted to point here that non-exitable exits stay in the queue like the exitable

```
struct Params {
        uint128 previousBondSize;
        uint128 updatedBondSize;
        uint128 previousExitBountySize;
        uint128 updatedExitBountySize;
        uint128 effectiveUpdateTime;
        uint16 lowerBoundDivisor;
        uint16 upperBoundMultiplier;
    }

```

`Params` is a structure for storing the bond size:

1. `previousBondSize` reflects the bond size, if `now` < `effectiveUpdateTime`.
2. `updatedBondSize` field reflects the bond size, when when the effectiveUpdateTime period has passed and `now` > `effectiveUpdateTime`.
3. `previousExitBountySize` reflects the bounty size, if `now` < `effectiveUpdateTime`.
4. The `updatedExitBountySize` field reflects the bounty size, when the effectiveUpdateTime period has passed and `now` > `effectiveUpdateTime`.
5. The `effectiveUpdateTime` field denotes the timestamp at which the updated bond and bounty size should reflect and be put to use. The value of the timestamp is 2 days from the last update timestamp.

> 6. `lowerBoundDivisor` is a factor which limits the bond to be updated to a certain minimum

Author: Does the Exit Bounty need an upper-bound and a lower-bound? or a free moving value within the Bond size should be accepted

Reviewer: I think any value from 0 to the bond size is fine.
We don't need to force it to e.g.  a certain percentage of the bond

# Exit Bond and Bounty Mechanics

This document is a description of the current implementation of exit bonds and process exit bounties which are an integral part of the exit mechanism.
### Terminology

> - **Exit Initiator** The person that starts an exit by either calling `startStandardExit()` or piggybacking on an in-flight exit

Reviewer 2: Nit pick, it's probably better to use `user` than `person` as the term is more generic.

- **Challenger** The person challenges a standard exit, an in-flight exit or a piggyback
- **Exit Processor** The person that calls `processExit()`
- **Exit Bond** The bond put up by the **Exit Initiator**

> - **Process Exit Bounty** A portion of the **Exit Bond** that is give to the **Exit Processor** as a reward for processing the exit.

Reviewer 2: ```suggestion
- **Process Exit Bounty** A portion of the **Exit Bond** that is given to the **Exit Processor** as a reward for processing the exit.
```

Whenever an exit is started, the **Exit Initiator** must put up an **Exit Bond**.

> If the exit is proven to be invalid the entire **Exit Bond** is awarded to the **Challenger**

Reviewer 2: ```suggestion
If the exit is proven to be invalid the entire **Exit Bond** is awarded to the **Challenger**.
```

If the exit is valid then after the challenge period it can be **processed** i.e. the funds returned to the **Exit Initiator**. A portion of the **Exit Bond**, called the **Process Exit Bounty**, is awarded to the **Exit Processor**. The remainder of the **Exit Bond** is returned to the **Exit Initiator**.
Note that the **Exit Initiator** can also be the **Exit Processor**.

## The Bond Life Cycle

### For Standard Exits

> The **Exit Bond** is supplied by the **Exit Initiator** when starting an exit. Its main motive is to disincentivize invalid exits, but in the case of a valid exit a portion of it is used to incentivize processing the exit by rewarding the **Exit Processor**.

Reviewer 3: ```suggestion
The **Exit Bond** is supplied by the **Exit Initiator** when starting an exit. Its main motivation is to disincentivize invalid exits, but in the case of a valid exit a portion of it is used to incentivize processing the exit by rewarding the **Exit Processor**.
```

The **Exit Initiator** should obtain the correct size of the Exit Bond by calling `startStandardExitBondSize()`.

**_After Starting an Exit:_**
*Bond -> from `Exit Initiator` to `Exit Game Contract`*

> For an invalid exit, a challenge will transfer the whole bond to the challenger. Non-exitable exits would still remain in the exit queue and is deleted when the exits are processed. For a scenario where there are mutliple invalid exits in the queue, the processor is still awarded with some gas reward for deleting non-exitable exits from storage.

Reviewer 2: ```suggestion
For an invalid exit, a challenge transfers the whole bond to the challenger. Non-exitable exits still remain in the exit queue and are deleted when the exits are processed. For a scenario where there are mutliple invalid exits in the queue, the processor is still awarded with some gas reward for deleting non-exitable exits from storage.
```

Reviewer 2: `For a scenario where there are mutliple invalid exits in the queue, the processor is still awarded with some gas reward for deleting non-exitable exits from storage.`

I case the exits are challenged successfully the exit processor does not get anything right?

Reviewer 2: https://github.com/omisego/plasma-contracts/blob/8468675915fdb202cf4656644621f7d4ebaea0da/plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessStandardExit.sol#L58-L63

Author: yes, except only for a small gas award from EVM for deleting storage

**_On succesfull challenge:_**
*Bond -> from `Exit Game Contract` to `Challenger`*

> The **Exit Processor** gets the bounty as a reward for processing an exit. The cost of processing multiple exits from the exit queue at once is less than the cost of processing all those exits individually. However, the **Process Exit Bounty** amount associated with each exit is given out to the **Exit Processor**.

Reviewer 2: ```suggestion
The **Exit Processor** gets the bounty as a reward for processing an exit. The cost of processing multiple exits from the exit queue at once is less than the cost of processing them individually. For simplicity and to save gas however, the **Process Exit Bounty** amount associated with each exit is given out to the **Exit Processor**.
```

> The **Exit Processor** gets the bounty as a reward for processing an exit. The cost of processing multiple exits from the exit queue at once is less than the cost of processing them individually. For simplicity and to save gas however, the **Process Exit Bounty** amount associated with each exit is given out to the **Exit Processor**.

Reviewer 3: > For simplicity and to save gas however,

I am a bit lost on what this part is aiming to explain actually 😅 where it is simplifying and saving gas?

Author: Thanks! I think it will be better to point out the stacking exits by the `exit processor` for more incentives part here instead, since saving gas is on their(the exit processors) side.

Author: updated `ddcb22f`

**_On succesfull challenge:_**
*Bond -> from `Exit Game Contract` to `Challenger`*

> The **Exit Processor** gets the bounty as a reward for processing an exit. The cost of processing multiple exits from the exit queue at once is less than the cumulative cost of processing them individually. Since, the **Process Exit Bounty** amount associated with each exit is given out, the **Exit Processor** can wait for the exit queue to stack up and then process exits together to get higher incentives.

Reviewer: ```suggestion
The **Exit Processor** gets the bounty as a reward for processing an exit. The cost of processing multiple exits from the exit queue at once is less than the cumulative cost of processing them individually. Since the **Process Exit Bounty** is given out for each exit processed, the **Exit Processor** can wait for the exit queue to stack up and then process multiple exits together to get a higher reward.
```
